src/rgw/rgw_tracer.h

```diff
@@ -0,0 +1,143 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab ft=cpp
+
+#ifndef RGW_TRACER_H
+#define RGW_TRACER_H
+
+#ifdef HAVE_JAEGER
+
+#define SIGNED_RIGHT_SHIFT_IS 1
+#define ARITHMETIC_RIGHT_SHIFT 1
+#include <jaegertracing/Tracer.h>
+
+#else
+
+#include <chrono>
+#include <string>
+#include <memory>
+#include <iostream>
+#include <initializer_list>
+#include <vector>
+
+namespace opentracing {
+inline namespace v3 {
+
+class StartSpanOption {};
+
+template <typename T> class option_wrapper {
+ public:
+  option_wrapper(const T &opt) {}
+};
+
+class SpanReference : public StartSpanOption {};
+
+class SpanContext {};
+
+class Span {
+  SpanContext ctx;
+ public:
+  Span() {}
+  virtual ~Span() {}
+  const SpanContext& context() { return ctx; }
+  template <typename T> void SetTag(std::string_view key, const T& value) const noexcept {}
+  void Log(std::initializer_list<std::pair<std::string_view, std::string_view>> fields) {}
+
+};
+
+class Tracer {
+ public:
+  Tracer() {}
+  ~Tracer() {}
+  std::unique_ptr<Span> StartSpan(std::string_view operation_name, std::initializer_list<option_wrapper<StartSpanOption>> option_list = {}) const {
+    return std::unique_ptr<Span>(new Span());
+}
```

Comment:
we shouldn't make stubs of the actual opentracing/jaeger types, especially when we have to allocate memory to use them. there shouldn't be any overhead when `HAVE_JAEGER` is disabled

the stub should be of our own `tracing::Tracer` class, and instead of returning `std::unique_ptr<opentracing::Span>`, we return some other value type that provides the same interface. for example, common/tracer.h has a `typedef std::unique_ptr<opentracing::Span> jspan;` - when `HAVE_JAEGER` is disabled, that could be a `class jspan` that overloads `operator->` and `operator*` like this:
```c++
#ifndef HAVE_JAEGER

// stub interface for opentracing::Span
struct span_stub {
  template <typename T>
  void SetTag(std::string_view key, const T& value) const noexcept {}

  void Log(std::initializer_list<std::pair<std::string_view, std::string_view>> fields) {}
};

// mimic interface of std::unique_ptr<Span> without using the heap
class jspan {
  span_stub span;
 public:
  span_stub& operator*() { return span; }
  const span_stub& operator*() const { return span; }

  span_stub* operator->() { return &span; }
  const span_stub* operator->() const { return &span; }
};

struct Tracer {
  bool is_enabled() const { return false; }
  jspan start_trace(std::string_view) { return {}; }
  jspan start_span(std::string_view, const jspan&) { return {}; }
};

#endif // !HAVE_JAEGER
```

Reply:
agree.

Reply:
done
@cbodley  do you think we also should create stubs for the tag keys such as `const auto BUCKET_NAME = "bucket_name";` 
to be something like `const auto BUCKET_NAME = "";`
which means once we want to add a new tag key, we'll need to add it twice.

Reply:
no, those strings don't have runtime overhead